app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt

```diff
@@ -1932,5 +1932,169 @@ class PatientRepositoryAndroidTest {
+        .inOrder()
+
+    assertThat(allPatientsInFacilityB)
+        .containsExactly(patientsInFacilityB[2].patient.uuid)
+        .inOrder()
+
+    assertThat(allPatientsInFacilityC).isEmpty()
+  }
+
+  @Test
+  fun querying_all_patients_in_a_facility_with_bps_in_different_facilities_should_set_the_last_bp_to_the_latest_one() {
+
+    fun recordPatientWithBps(
+        patientUuid: UUID,
+        patientName: String,
+        recordBpsAt: List<Triple<UUID, Facility, Instant>>
+    ): PatientProfile {
+      val patientProfile = testData.patientProfile(patientUuid = patientUuid).let { patientProfile ->
+        patientProfile.copy(patient = patientProfile.patient.copy(status = PatientStatus.ACTIVE, fullName = patientName))
+      }
+
+      val bps = recordBpsAt.map { (uuid, facility, recordedAt) ->
+        testData.bloodPressureMeasurement(uuid = uuid, patientUuid = patientUuid, facilityUuid = facility.uuid, recordedAt = recordedAt)
+      }
+
+      patientRepository.save(listOf(patientProfile)).blockingAwait()
+      bloodPressureRepository.save(bps).blockingAwait()
+
+      return patientProfile
+    }
+
+    fun allPatientIdsAndLastBpRecordedAtFacility(facility: Facility): List<Pair<UUID, Instant>> {
+      return patientRepository.allPatientsInFacility(facility).blockingFirst().map { it.uuid to it.lastBp!!.takenOn }
+    }
+
+    // given
+    val (facilityA, facilityB) = facilityRepository.facilities().blockingFirst()
+    val now = Instant.parse("2019-06-26T00:00:00Z")
+
+    val bpUuid1ForPatientA = UUID.fromString("588b0d8c-6dce-4664-863f-8c79c06c6c3c")
+    val bpUuid2ForPatientA = UUID.fromString("ba875b98-4ca0-4097-bbc7-56041b377cf2")
+    val bpUuid3ForPatientA = UUID.fromString("4b92601e-4a86-40de-b6b7-707f9122a81b")
+
+    val patientWithLatestBpInFacilityA = recordPatientWithBps(
+        UUID.fromString("e2a49529-3b46-4c9e-a7bc-af3090f1ecbb"),
+        "Patient A",
+        listOf(
+            Triple(bpUuid1ForPatientA, facilityA, now.plusSeconds(1)),
+            Triple(bpUuid2ForPatientA, facilityB, now.minusSeconds(1)),
+            Triple(bpUuid3ForPatientA, facilityA, now)
+        )
+    )
+
+    val bpUuid1ForPatientB = UUID.fromString("00e39456-f5e4-4538-956d-7e973ec5da88")
+    val bpUuid2ForPatientB = UUID.fromString("b72e28cd-1b46-4c72-84cd-58530c2829e3")
+    val bpUuid3ForPatientB = UUID.fromString("c18b41d5-260e-4a26-8030-45c621ded98d")
+
+    val patientWithLatestBpInFacilityB = recordPatientWithBps(
+        UUID.fromString("7925e13f-3b04-46b0-b685-7005ebb1b6fd"),
+        "Patient B",
+        listOf(
+            Triple(bpUuid1ForPatientB, facilityB, now.plusSeconds(5)),
+            Triple(bpUuid2ForPatientB, facilityA, now.minusSeconds(1)),
+            Triple(bpUuid3ForPatientB, facilityB, now)
+        )
+    )
+
+    // when
+    val allPatientsInFacilityA = allPatientIdsAndLastBpRecordedAtFacility(facilityA)
```

Comment:
Why do we need to create a `Pair` here?
Again this makes it difficult to read and understand what we want to verify.
In the assertions, can't we directly verify ids and bp recordedAt for `allPatientsInFacility`

Reply:
We could. I'll look at making this test more understandable.

Reply:
Take a look now? I've rewritten the test to make it (hopefully) more understandable.